designs/2019-suggestions/README.md

```diff
@@ -0,0 +1,59 @@
+
+## Motivation
+
+This has been suggested multiple times before, but the main driving force behind this proposal is ability to provide user with options on how to automatically refactor/fix their code, without a possibility of "stealthily" breaking it. Examples of such suggestions include ability to suggest adding `import` statement if certain code is added to a file by the user, ability to provide refactoring suggestions like extracting repeated value into a variable, reformatting code to extract repeated statements into reusable function, etc. I do not envision ESLint team to initially create any rules that would provide suggestions, so this would be a feature that would enable plugin authors to do so, but it might be something that ESLint team will eventually do. This feature is geared towards code editor integrations, as such, it will only be available via NodeJS API.
+
+## Detailed Design
+
+* Modify `context.report` function parameter to include new property called `suggest` of type array of functions. Each function will be able to modify the code of the file to implement suggestion. Syntax is going to be exactly the same as for `fix` functions. Example of new signature below:
+```js
+context.report({
+    node,
+    loc,
+    messageId: 'messageId',
+    fix(fixer) { },
+    suggest: [
+        (fixer) => { },
+        (fixer) => { }
+    ]
+});
+```
+* Change `_verifyWithoutProcessors` and `_verifyWithProcessor` functions in linter.js to modify output of those functions to include `suggestions` array returned by the rule.
+* Modify output of `CLIEngine` `executeOnFiles` and `executeOnText` to include `suggestions` array returned by the rule.
+* Introduce a new `CLIEngine` public function `outputSuggestion(errorReport, suggestion)` that would apply selected suggestion (second argument) to the file that the errorReport was created for and save it to disk.
+
+## Documentation
+
+This would need to be documented as part of the public API.
+* [Node.js API](https://eslint.org/docs/developer-guide/nodejs-api) page needs to be updated with corresponding changes to `linter` and `CLIEngine`.
+* [Working with Rules](https://eslint.org/docs/developer-guide/working-with-rules) page needs to be updated to show examples of returning suggestions.
+
+## Drawbacks
+
+This might potentially be a bit confusing, since we do not have many functions that we provide through API only and not through CLI.
+
+## Backwards Compatibility Analysis
+
+Since `suggestions` property in `context.report` object is optional, this should be fully backwards compatible.
+
+## Alternatives
+
+This functionality is basically the same as existing `autofix` functionality already included in ESLint. The only difference is there might be multiple suggestions per error and they are not applied automatically by ESLint.
+
+## Open Questions
+
+Should we consider instead providing new type of `rules` that would not report any issues, but only provide suggestions?
```

Comment:
A new severity might be useful.

Severity | CLI shows | CLI fails | Editors can use
:--------|:----------|:----------|:-----------
off      | ✖        | ✖        | ✖
option  | ✖        | ✖        | ✔
warn     | ✔        | ✖        | ✔
error    | ✔        | ✔        | ✔

※ The human-readable formatters such as `stylish` doesn't show the new "option" severity. This is the difference from "warn" severity.

Reply:
I would like to avoid doing that as part of this RFC. I think it's fine to require rule to output an error/warning for now. If we find later that there are a lot of community rules that want to provide suggestions without reporting actual errors, we can always introduce new level later (but I think that might actually create more confusion for users, since `error`/`warning` is user defined level, and `option` would be more of a rule defined one).